gateway/src/serial_console.rs

```diff
@@ -4,260 +4,150 @@
-    });
-
-    // `.body()` only fails if our headers are bad, which they aren't
-    // (unless `hyper::handshake` gives us a bogus accept key?), so we're
-    // safe to unwrap this
-    Ok(http::Response::builder()
-        .status(http::StatusCode::SWITCHING_PROTOCOLS)
-        .header(header::CONNECTION, "Upgrade")
-        .header(header::UPGRADE, "websocket")
-        .header(header::SEC_WEBSOCKET_ACCEPT, accept_key)
-        .body(Body::empty())
-        .unwrap())
-}
-
 #[derive(Debug, thiserror::Error)]
 enum SerialTaskError {
     #[error(transparent)]
-    Error(#[from] Error),
+    SpCommsError(#[from] SpCommsError),
     #[error(transparent)]
     TungsteniteError(#[from] tokio_tungstenite::tungstenite::Error),
 }
 
-struct SerialConsoleTask {
+pub(crate) async fn run(
     console: AttachedSerialConsole,
-    ws_stream: WebSocketStream<Upgraded>,
-}
-
-impl SerialConsoleTask {
-    async fn run(self, log: &Logger) -> Result<(), SerialTaskError> {
-        let (ws_sink, ws_stream) = self.ws_stream.split();
-
-        // Spawn a task to send any messages received from the SP to the client
-        // websocket.
-        //
-        // TODO-cleanup We have no way to apply backpressure to the SP, and are
-        // willing to buffer up an arbitray amount of data in memory. We should
-        // apply some form of backpressure (which the SP could only handle by
-        // discarding data).
-        let (ws_sink_tx, ws_sink_rx) = mpsc::unbounded_channel();
-        let mut ws_sink_handle =
-            tokio::spawn(Self::ws_sink_task(ws_sink, ws_sink_rx));
-
-        // Spawn a task to send any messages received from the client websocket
-        // to the SP.
-        let (console_tx, mut console_rx) = self.console.split();
-        let console_tx = DetachOnDrop::new(console_tx);
-        let mut ws_recv_handle = tokio::spawn(Self::ws_recv_task(
-            ws_stream,
-            console_tx,
-            log.clone(),
-        ));
-
-        loop {
-            tokio::select! {
-                // Our ws_sink task completed; this is only possible if it
-                // fails, since it loops until we drop ws_sink_tx (which doesn't
-                // happen until we return!).
-                join_result = &mut ws_sink_handle => {
-                    let result = join_result.expect("ws sink task panicked");
-                    return result;
-                }
+    conn: WebsocketConnection,
+    log: Logger,
+) -> WebsocketChannelResult {
+    let upgraded = conn.into_inner();
+    let config =
+        WebSocketConfig { max_send_queue: Some(4096), ..Default::default() };
+    let ws_stream =
+        WebSocketStream::from_raw_socket(upgraded, Role::Server, Some(config))
+            .await;
+
+    let (ws_sink, ws_stream) = ws_stream.split();
+
+    // Spawn a task to send any messages received from the SP to the client
+    // websocket.
+    //
+    // TODO-cleanup We have no way to apply backpressure to the SP, and are
```

Comment:
Would it be easier, and perhaps just as useful to use a bounded channel here and drop data directly, rather than tell the SP to drop data?

Reply:
(Just for clarity - this is not new code and is unchanged by this PR)

We could, yeah. I'm not sure how to pick a bound for this channel - the unit we're sending on it is "packets from the SP" which have somewhere between 1 and ~1024 bytes in them, so maybe it's fine to just pick a pretty big bound?

Reply:
Yeah, I'd set it to just something quite large. Even 1MB means we have 1000 max sized packets queued up. So 10MB probably good enough.

Reply:
Set it to 10,000 and added a bit of logging in 87124c2. Chipping away at those TODOs, thanks!